commit b20d572e5661efa319ec7f4c08629be16d6b4aa9
Author: Tom Sommer <tomsommer@users.noreply.github.com>
Date:   Thu Jun 13 12:17:55 2019 +0200

    Fix type conversions
    
    Calling SvNV() for magical scalar is not enough for float type conversion.
    It caused problem for Amavis in tainted mode -- all float values were zero.
    On the other hand SvIV() and SvUV() seems to work fine. To be sure that
    correct value of float is in scalar use sv_setnv() with explicit NV float
    value. Similar code is changed also for integers IV/UV.
    
    Fixes #78, #312
    
    Credit https://github.com/kentnl-gentoo/DBD-mysql/commit/b6b8540216bd03b68f3bc076b3d3106f4be23f9d

diff --git a/dbdimp.c b/dbdimp.c
index 39753c8..717ed8e 100644
--- a/dbdimp.c
+++ b/dbdimp.c
@@ -4448,8 +4448,7 @@ process:
           if (!(fields[i].flags & ZEROFILL_FLAG))
           {
             /* Coerce to double and set scalar as NV */
-            (void) SvNV(sv);
-            SvNOK_only(sv);
+            sv_setnv(sv, SvNV(sv));
           }
           break;
 
@@ -4460,13 +4459,11 @@ process:
             /* Coerce to integer and set scalar as UV resp. IV */
             if (fields[i].flags & UNSIGNED_FLAG)
             {
-              (void) SvUV(sv);
-              SvIOK_only_UV(sv);
+              sv_setuv(sv, SvUV(sv));
             }
             else
             {
-              (void) SvIV(sv);
-              SvIOK_only(sv);
+              sv_setiv(sv, SvIV(sv));
             }
           }
           break;
